checker/src/org/checkerframework/checker/lowerbound/LowerBoundAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,653 @@
+            }
+            return;
+        }
+
+        /**
+         *  Determine the annotation that should be associated with a literal.
+         */
+        private AnnotationMirror anmFromVal(int val) {
+            if (val == -1) {
+                return GTEN1;
+            } else if (val == 0) {
+                return NN;
+            } else if (val > 0) {
+                return POS;
+            } else {
+                return UNKNOWN;
+            }
+        }
+
+        @Override
+        public Void visitLiteral(LiteralTree tree, AnnotatedTypeMirror type) {
+            // We could call the constant Value Checker here but if we already know it's a literal...
+            if (tree.getKind() == Tree.Kind.INT_LITERAL
+                    || tree.getKind() == Tree.Kind.LONG_LITERAL
+                    || tree.getKind() == Tree.Kind.CHAR_LITERAL) {
+                int val = (int) tree.getValue();
+                type.addAnnotation(anmFromVal(val));
+            }
+            return super.visitLiteral(tree, type);
+        }
+
+        /**
+         *  Call increment and decrement helper functions.
+         */
+        @Override
+        public Void visitUnary(UnaryTree tree, AnnotatedTypeMirror typeDst) {
+            AnnotatedTypeMirror typeSrc = getAnnotatedType(tree.getExpression());
+            switch (tree.getKind()) {
+                case PREFIX_INCREMENT:
+                    promoteType(typeSrc, typeDst);
+                    break;
+                case PREFIX_DECREMENT:
+                    demoteType(typeSrc, typeDst);
+                    break;
+                case POSTFIX_INCREMENT: // Do nothing. The hack above takes care of these.
+                    break;
+                case POSTFIX_DECREMENT:
+                    break;
+                default:
+                    break;
+            }
+            return super.visitUnary(tree, typeDst);
+        }
+
+        /**
+         *  Get the list of possible values from a Value Checker type.
+         *  May return null.
+         */
+        private List<Long> possibleValuesFromValueType(AnnotatedTypeMirror valueType) {
+            AnnotationMirror anm = valueType.getAnnotation(IntVal.class);
+            // Anm can be null if the Value Checker didn't assign an IntVal annotation
+            if (anm == null) {
+                return null;
+            }
+            return ValueAnnotatedTypeFactory.getIntValues(anm);
+        }
+
+        /**
+         * If the argument valueType indicates that the Constant Value
+         * Checker knows the exact value of the annotated expression,
+         * returns that integer.  Otherwise returns null.
+         */
+        public Integer maybeValFromValueType(AnnotatedTypeMirror valueType) {
+            List<Long> possibleValues = possibleValuesFromValueType(valueType);
+            if (possibleValues != null && possibleValues.size() == 1) {
+                return new Integer(possibleValues.get(0).intValue());
```

Comment:
Should this return the min if `possibleValues.size() >0`?  (If so, maybe change the name to minValFromValueType.)